services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartS3AsyncClient.java

```diff
@@ -17,31 +17,59 @@
 
 
 import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
 import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.ApiName;
 import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.core.async.AsyncResponseTransformer;
 import software.amazon.awssdk.services.s3.DelegatingS3AsyncClient;
 import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.internal.UserAgentUtils;
 import software.amazon.awssdk.services.s3.model.CopyObjectRequest;
 import software.amazon.awssdk.services.s3.model.CopyObjectResponse;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
 import software.amazon.awssdk.services.s3.model.PutObjectRequest;
 import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+import software.amazon.awssdk.services.s3.model.S3Request;
+import software.amazon.awssdk.services.s3.multipart.MultipartConfiguration;
+import software.amazon.awssdk.utils.Validate;
 
-// This is just a temporary class for testing
-//TODO: change this
+/**
+ * An {@link S3AsyncClient} that automatically converts put, copy requests to their respective multipart call. Note: get is not
+ * yet supported.
+ *
+ * @see MultipartConfiguration
+ */
 @SdkInternalApi
-public class MultipartS3AsyncClient extends DelegatingS3AsyncClient {
-    private static final long DEFAULT_PART_SIZE_IN_BYTES = 8L * 1024 * 1024;
+public final class MultipartS3AsyncClient extends DelegatingS3AsyncClient {
+
+    private static final ApiName USER_AGENT_API_NAME = ApiName.builder().name("hll").version("s3Multipart").build();
+
+    private static final long DEFAULT_MIN_PART_SIZE = 8L * 1024 * 1024;
     private static final long DEFAULT_THRESHOLD = 8L * 1024 * 1024;
+    private static final long DEFAULT_API_CALL_BUFFER_SIZE = DEFAULT_MIN_PART_SIZE * 2;
```

Comment:
Based on the perf result, should we bump this to `* 4` or keep it as is?

Reply:
Yup, I ran some test and going to 4x seems to be a good balance. It is an improvement that seems to be worth it, but going higher sees diminishing returns:

## 4GiB Upload Result - 8 MiB part size
| api call buffer size | Average latency (s) | Average throughput (Gbps) |
|---|---|---|
| 2x (16 Mib) |  6.359 | 5.032 |
| 3x (24 Mib) |  6.598 | 4.850 |
| 4x (32 Mib) | 5.585 | 5.729 |
| 8x (64 Mib) |  5.760 | 5.556 |
| 3 GiB | 4.700 | 6.808 |